platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,279 @@
+// Copyright 2022 Marek Kraus (@gamelaster)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+
+#include "hardware/pio.h"
+#include "hardware/clocks.h"
+#include "ps2.h"
+#include "print.h"
+
+#if !defined(MCU_RP)
+#    error PIO Driver is only available for Raspberry Pi 2040 MCUs!
+#endif
+
+#if defined(PS2_ENABLE)
+#    if defined(PS2_MOUSE_ENABLE)
+#        if !defined(PS2_MOUSE_USE_REMOTE_MODE)
+#            define BUFFERED_MODE_ENABLE
+#        endif
+#    else // PS2 Keyboard
+#        define BUFFERED_MODE_ENABLE
+#    endif
+#endif
+
+#if PS2_DATA_PIN + 1 != PS2_CLOCK_PIN
+#    error PS/2 Clock pin must be followed by data pin!
+#endif
+
+static inline void pio_serve_interrupt(void);
+
+#if defined(PS2_PIO_USE_PIO1)
+static const PIO pio = pio1;
+
+OSAL_IRQ_HANDLER(RP_PIO1_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#else
+static const PIO pio = pio0;
+
+OSAL_IRQ_HANDLER(RP_PIO0_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#endif
+
+#define PS2_WRAP_TARGET 0
+#define PS2_WRAP 28
+
+// clang-format off
+static const uint16_t ps2_program_instructions[] = {
+            //     .wrap_target
+    0xe020, //  0: set    x, 0
+    0x8080, //  1: pull   noblock
+    0xa027, //  2: mov    x, osr
+    0x0046, //  3: jmp    x--, 6
+    0x00c0, //  4: jmp    pin, 0
+    0x0015, //  5: jmp    21
+    0xe082, //  6: set    pindirs, 2
+    0xef00, //  7: set    pins, 0                [15]
+    0xaf42, //  8: nop                           [15]
+    0xe283, //  9: set    pindirs, 3             [2]
+    0xe081, // 10: set    pindirs, 1
+    0xe002, // 11: set    pins, 2
+    0xe029, // 12: set    x, 9
+    0x2021, // 13: wait   0 pin, 1
+    0x6601, // 14: out    pins, 1                [6]
```

Comment:
Ideally you should somehow keep the open drain behavior of the data pin instead of driving it high actively, as may happen here.  Maybe something like `out pindirs, 1` would work here (but you would need to invert the data somewhere).